lib/host/gen_helpers.js

```diff
@@ -219,7 +219,7 @@ const genHelpers = (state = {}, adapter) => {
       // Note that we don't wait for the order to submit here, since it might
       // fill immediately triggering a fill event before we patch the state
       // orders/allOrders objects
-      adapter.submitOrderWithDelay(connection, delay, order).then((order) => { // eslint-disable-line
+      await adapter.submitOrderWithDelay(connection, delay, order).then((order) => { // eslint-disable-line
```

Comment:
we can't await here.. because as a result of order submission,  we receive different events like `order_new`, `order_fill` and `order_cancel`  which is being handled here `https://github.com/bitfinexcom/bfx-hf-algo/blob/master/lib/host/ws2/process_message.js` to update the order statuses in state which is required for algo orders to process them accordingly

Reply:
@avsek477  this is necessary to make sure that the request has been performed, otherwise it will always be considered completed even if there is no Internet connection

Reply:
hm, that would block the whole execution path, indeed we can't wait here...

people can currently define a "submit delay". its measured in seconds. when you define a submit delay of 30 seconds the promise won't resolve for that timeframe

Reply:
i agree with @vitaliystoliarovcc, too

where we need parallelism in order handling, we could use things like async.parallel / promise.all and keep a map of current open orders. the current sequential emulation we have is too limited, but also has its uses for simple algos

Reply:
suggestion: we remove the "feature" submitDelay - similar like we did remove cancel-delays in the last release. the feature has no practical use and complicates the code base. after that is done, we add concurrent handling to the algo orders that need it. for example twap orders, there is just one open order, so we can go with "awaiting" a successful order submit.

ping/pong orders support multiple order submissions at a time. for that the function should take an array of orders and resolve only when all orders are submitted (in case some are cancelled e.g. because of invalid balance it should also fullfill the promise, am i am thinking correct here?)

so two separate tasks:

1. remove submitDelay in all algo orders, submit one PR per algo order, we can hardcode submit delay to 0 in this intermediate step
2. change the submit order function to: not having any delay functionality, accept multiple orders for bulk submits

this way we solve a few issues at once we have with order submits and can also resolve the "no-wifi" issue. does the plan sound good to you @avsek477 @vitaliystoliarovcc ?

Reply:
> similar like we did remove cancel-delays

can you share the link to the PR?

Reply:
https://github.com/bitfinexcom/bfx-hf-algo/pull/90
https://github.com/bitfinexcom/bfx-hf-algo/pull/91
https://github.com/bitfinexcom/bfx-hf-algo/pull/92
https://github.com/bitfinexcom/bfx-hf-algo/pull/93
https://github.com/bitfinexcom/bfx-hf-algo/pull/94
https://github.com/bitfinexcom/bfx-hf-algo/pull/95

Reply:
> (in case some are cancelled e.g. because of invalid balance it should also fullfill the promise, am i am thinking correct here?)

Makes sense to me


As I understood, the main problem with "await" is the inclusion of a delay in the submitOrderWithDelay helper, so using a loop, each next order will wait for all previous ones.

For example, the app emits
```
emit('exec:order:submit:all', gid, [A,B,C], submitDelay)
```
where A, B, C - orders

1. Without _await_ (changes in `master` branch - NOT OK, the submit responses are ignored)
```
A->++++++>==>
   B->++++++>=========>X - Ooops, request timeout
      C->++++++>==> 

/// Interval's promise:
--------> // Promise resolved despite the request is still running
```

2. With _await_ (this PR's changes - NOT OK, each next order in list will wait for the delay of the previous one)

```
A->++++++>==>
             B->++++++>=========>X - Ooops, request timeout
                                 C->++++++>==> 

/// Interval's promise:
---------------------------------------------> // Promise resolved when all request are fulfilled, but it takes to long if we submit more than one order
```

3. With _await_ and Promise.all (TODO)

```
A->++++++>==>
B->++++++>=========>X - Ooops, request timeout
C->++++++>==> 

/// Interval's promise:
----------------------> // Promise resolved in time equal to the longest order submission
```

Legend: 

-> - Promise resolving
++++> = Delay
==> Request

@robertkowalski @avsek477 is it correct?

The simplest solution I see is to replace a loop with Promise.all in [submit_all_orders.js](https://github.com/bitfinexcom/bfx-hf-algo/blob/master/lib/host/events/submit_all_orders.js)

What do you think?

Reply:
> we remove the "feature" submitDelay
> remove submitDelay in all algo orders, submit one PR per algo order, we can hardcode submit delay to 0 in this intermediate step

But if we remove `submitDelay`, it will affect some orders:
MA Crossover, OcO, Ping/Pong and Accumulate/Distribute

"Delay" can be moved next to `emit('exec:order:submit:all')`, as it was done for Iceberg

https://github.com/bitfinexcom/bfx-hf-algo/blob/405abcbb8b10c0686246ab707590c8834bad0266/lib/iceberg/events/self_submit_orders.js#L22-L29

But will this work for all types of Algo orders? Perhaps it will be important to perform some validation (inside `submitOrderWithDelay` in bfx-hf-algo/lib/host/gen_helpers.js and bfx-hf-algo/lib/ws_adapter.js) before delay

Reply:
> With await and Promise.all (TODO)
> @robertkowalski @avsek477 is it correct?

yes this is correct

> But will this work for all types of Algo orders? Perhaps it will be important to perform some validation (inside submitOrderWithDelay in bfx-hf-algo/lib/host/gen_helpers.js and bfx-hf-algo/lib/ws_adapter.js) before delay

all algo orders will need to get touched and then tested. best to start with TWAP for now and submit a PR for it. algo orders behaviour need to be studied carefully, they can be tricky sometimes. ping/pong might need more in-depth changes to the acutal code, when you try algo orders currently you can select an order count > 1, in those cases many orders will be sent to the server concurrently and stay open in parallel. so each algo must be tested carefully before making the changes

Reply:
> best to start with TWAP for now and submit a PR for it.

if I make changes in submit_all_orders.js, all order types will be affected

Reply:
> > best to start with TWAP for now and submit a PR for it.
> 
> if I make changes in submit_all_orders.js, all order types will be affected

We can create a new event similar to `exec:order:submit:all` in `ao_host.js` and use the new event for submission of orders for twap orders for now